{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "70792dd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Packages\n",
    "\n",
    "import re\n",
    "import pandas as pd\n",
    "from Bio import Entrez, SeqIO, SeqFeature\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "# Email for Entrez\n",
    "Entrez.email ='your email'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9783a1f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open Twist order sheet as DataFrame\n",
    "\n",
    "fasta_df = pd.read_excel('20240408_Twist_Order.xlsx')\n",
    "\n",
    "# Remove FASTA formating of '>' character in 'Accession' column\n",
    "\n",
    "fasta_df['Accession'] = fasta_df['Accession'].str[1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b015158a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def downstream_sequence(row):\n",
    "    \n",
    "    '''\n",
    "    This function takes Rfam entry information and outputs the coding gene downstream the entry.\n",
    "    \n",
    "    Input: Row from DataFrame FASTA file of \n",
    "        species' NCBI accession ID,\n",
    "        genomic positions,\n",
    "        Rfam entry sequence\n",
    "    \n",
    "    Output:\n",
    "        codon_start = the location of the start position of the coding gene (provided by access_NCBI)\n",
    "        gene = name of the coding gene (provided by access_NCBI)\n",
    "    '''\n",
    "    \n",
    "    # Initialize values\n",
    "    \n",
    "    gene = int(0)\n",
    "    x = 50 # Iterate over 50 nts looking for coding region\n",
    "    \n",
    "    # Rfam entry on the Positive strand:\n",
    "    \n",
    "    if row['Start'] < row['Stop']: \n",
    "        \n",
    "        # Incimentially search further downstream the aptamer until hit coding region\n",
    "        \n",
    "        while type(gene) != str:\n",
    "            \n",
    "            # Establish positions in the genome\n",
    "            \n",
    "            start = row['Start']\n",
    "            stop = row['Stop'] + x\n",
    "            \n",
    "            # Send genomic position to NCBI to identify downstream gene\n",
    "            \n",
    "            codon_start, gene = access_NCBI(start,stop, x) # gene returns a int(0) until coding region is found, which stops the loop\n",
    "                \n",
    "            # Check the next 50 nts for a protein retion    \n",
    "                \n",
    "            x = x + 50 \n",
    "            \n",
    "            if x > 1049: # Give up if no coding region is found within 1000 nts after the Rfam entry\n",
    "                gene = 'Further than 1000 nts'\n",
    "     \n",
    "    # Rfam entry on the Negative strand:\n",
    "    \n",
    "    else:\n",
    "\n",
    "        # Incimentially search further downstream the aptamer until hit protein\n",
    "        \n",
    "        while type(gene) != str :\n",
    "            \n",
    "            # Establish positions in the genome\n",
    "            \n",
    "            start = row['Stop'] - x\n",
    "            stop = row['Start']\n",
    "            \n",
    "            # Send genomic position to NCBI to identify downstream gene\n",
    "            \n",
    "            codon_start, gene = access_NCBI(start,stop, x) # gene returns a int(0) until coding region is found, which stops the loop\n",
    "                \n",
    "            # Check the next 50 nts for a protein retion    \n",
    "                \n",
    "            x = x + 50\n",
    "            \n",
    "            if x > 1049: # Give up if no coding region is found within 1000 nts after the Rfam entry\n",
    "                gene = 'Further than 1000 nts'\n",
    "    \n",
    "    return codon_start, gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e684dc53",
   "metadata": {},
   "outputs": [],
   "source": [
    "def access_NCBI(start, stop, x):\n",
    "    \"\"\"\n",
    "    This function takes genomic positions and provides if there is a coding region and what it is.\n",
    "    \n",
    "    Input: \n",
    "        start = genome start position\n",
    "        stop = genome stop position\n",
    "        x = distance into the coding region to back track to start codon\n",
    "    \n",
    "    Output:\n",
    "        codon_start = the location of the start position of the coding gene\n",
    "        gene = name of the coding gene\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # Collect genome region information from NCBI\n",
    "            \n",
    "    handle = Entrez.efetch(db=\"nucleotide\", \n",
    "                           id=row['Accession'], \n",
    "                           rettype=\"gb\", \n",
    "                           retmode=\"text\", \n",
    "                           seq_start= start, \n",
    "                           seq_stop= stop)\n",
    "            \n",
    "    # Save genomic information\n",
    "\n",
    "    GenomeSeq = SeqIO.read(handle, \"genbank\")\n",
    "\n",
    "    for feature in GenomeSeq.features:\n",
    "\n",
    "        # Identify CDS in genomic information\n",
    "        if feature.type == \"CDS\":  \n",
    "            location = feature.location\n",
    "            if int(location.start) > 0:\n",
    "                codon_start = int(location.start) - aptamer_length + 1\n",
    "            else:\n",
    "                codon_location = int(location.end)\n",
    "                codon_start = int(x) - int(location.end)\n",
    "\n",
    "            try:\n",
    "                product = feature.qualifiers['product']\n",
    "                gene = product[0]\n",
    "\n",
    "            # If there are no CDS in the genomic information, the list is appended with 'unknown'\n",
    "\n",
    "            except KeyError:\n",
    "                gene = 'unknown'\n",
    "    \n",
    "            return codon_start, str(gene)\n",
    "    \n",
    "    return 'tbd' , int(0) # Loop continues until a coding region is found or x > 1049 nts from the entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "dee0a13d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initalize lists\n",
    "\n",
    "aptamer_length_list = []\n",
    "strand_list = []\n",
    "codon_start_list = []\n",
    "aptamer_downstream_gene_list = []\n",
    "\n",
    "# Calculate entry length\n",
    "\n",
    "aptamer_length = len(row['Sequence'])\n",
    "\n",
    "# Go through each DataFrame row\n",
    "\n",
    "for index, row in fasta_df.iterrows():\n",
    "    \n",
    "    # Run functions to get coding information after Rfam entry\n",
    "    \n",
    "    aptamer_length, codon_start, aptamer_downstream_gene = downstream_sequence(row)\n",
    "    \n",
    "    # Save information into list\n",
    "    \n",
    "    aptamer_length_list.append(aptamer_length)\n",
    "    codon_start_list.append(codon_start)\n",
    "    aptamer_downstream_gene_list.append(aptamer_downstream_gene)\n",
    "\n",
    "# Add infromation from lists to DataFrame\n",
    "\n",
    "fasta_df['Aptamer Length'] = aptamer_length_list\n",
    "fasta_df['Nts between Aptamer and Start Codon'] = codon_start_list\n",
    "fasta_df['Gene'] = aptamer_downstream_gene_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed04e1aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export DataFrame to Excel\n",
    "\n",
    "fasta_df.to_excel('Fluoride_CDS.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "372f146d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
