{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c16447f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# install packages\n",
    "\n",
    "import re\n",
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqFeature\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "#use own email\n",
    "Entrez.email ='Your Email'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "92d95696",
   "metadata": {},
   "outputs": [],
   "source": [
    "# J23119 E. coli promoter\n",
    "\n",
    "five_seq = 'gcttccggcttgattctaaagatctttgacagctagctcagtcctaggtataatactagt'\n",
    "\n",
    "# 3` primer binding site\n",
    "\n",
    "three_seq = 'gggcacaaattttctgtccg'\n",
    "\n",
    "# Oligo synethesis length maximum\n",
    "\n",
    "TWIST_max_nt_synthsis = 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2541ec6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the file path to the FASTA dowloaded from Rfam\n",
    "file_path = '20230524_RF01734.fa'\n",
    "\n",
    "file = open('20230524_RF01734.fa','r')\n",
    "dna_seq = file.readlines()\n",
    "file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8d01f0a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize lists\n",
    "\n",
    "ID = []\n",
    "Sequence = []\n",
    "Start = []\n",
    "Stop = []\n",
    "\n",
    "# Organize Rfam FASTA file into DataFrame\n",
    "\n",
    "for line in dna_seq:\n",
    "    if re.search('>', line) != None: # If the line is the Accession code\n",
    "        line = line.split('/')\n",
    "        ID.append(line[0])\n",
    "        genome_start = line[1].split('-')\n",
    "        Start.append(genome_start[0])\n",
    "        genome_stop = genome_start[1].split(' ')\n",
    "        Stop.append(genome_stop[0])\n",
    "    else:\n",
    "        Sequence.append(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "de7c2a3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create DataFrame\n",
    "\n",
    "fasta_df = pd.DataFrame({'Accession': ID, 'Start': Start, 'Stop': Stop, 'Sequence': Sequence})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b3d8ba4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def access_NCBI(accession, start, stop, strand, aptmer_length):\n",
    "    \"\"\"\n",
    "    This function takes genomic positions and saves the nucleotide sequence of the next 150 nucleotides downstream the region stop\n",
    "    \n",
    "    Input: start and stop positions within a genome\n",
    "    \n",
    "    Output:\n",
    "        sequence = full sequence from RFam next 150 nts downstream\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    extend_length = TWIST_max_nt_synthsis - (len(five_seq)+\n",
    "                                              len(three_seq)+\n",
    "                                             aptamer_length) - 1\n",
    "    if strand == 'Positive':\n",
    "        stop = int(stop) + extend_length\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", \n",
    "                               id=accession, \n",
    "                               rettype=\"gb\", \n",
    "                               retmode=\"text\", \n",
    "                               seq_start=start, \n",
    "                               seq_stop=stop)\n",
    "    \n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "        sequence = record.seq\n",
    "        handle.close()\n",
    "    \n",
    "    else: # For scaffolds on the Negative Strand\n",
    "        start = int(start) - extend_length\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", \n",
    "                                       id=accession, \n",
    "                                       rettype=\"gb\", \n",
    "                                       retmode=\"text\", \n",
    "                                       seq_start=start, \n",
    "                                       seq_stop=stop)\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "        sequence = record.seq\n",
    "        handle.close()\n",
    "        sequence = sequence.reverse_complement()\n",
    "    \n",
    "    return str(sequence), extend_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "55d551a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define lists\n",
    "nucleotides_list = []\n",
    "aptamer_length_list = []\n",
    "strand_list = []\n",
    "codon_start_list = []\n",
    "aptamer_downstream_gene_list = []\n",
    "nt_len = []\n",
    "extended_list = []\n",
    "\n",
    "# Run NCBI information collection for each row\n",
    "\n",
    "for index, row in fasta_df.iterrows():\n",
    "    accession = row['Accession'][1:]\n",
    "    if re.search(r'^U', accession) == None:\n",
    "    \n",
    "        if int(row['Start']) < int(row['Stop']):\n",
    "            start = row['Start']\n",
    "            stop = row['Stop']\n",
    "            strand = 'Positive'\n",
    "        else:\n",
    "            start = row['Stop']\n",
    "            stop = row['Start']\n",
    "            strand = 'Negative'\n",
    "        \n",
    "        aptamer_length = int(stop)-int(start)\n",
    "        \n",
    "# Follow along with the search\n",
    "\n",
    "        print('Currently looking for: '+ accession+' with aptamer length of '+str(aptamer_length))\n",
    "\n",
    "        nucleotides, extended = access_NCBI(accession, start, stop, strand, aptamer_length)\n",
    "\n",
    "        # Save information into list\n",
    "\n",
    "        aptamer_length_list.append(int(stop) - int(start))\n",
    "        nucleotides_list.append(nucleotides)\n",
    "        strand_list.append(strand)\n",
    "        nt_len.append(len(nucleotides))\n",
    "        extended_list.append(extended)\n",
    "    else:\n",
    "        aptamer_length_list.append('tbd')\n",
    "        nucleotides_list.append('tbd')\n",
    "        strand_list.append('tbd')\n",
    "        nt_len.append('tbd')\n",
    "        extended_list.append('tbd')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dac902a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add gathered information to DataFrame\n",
    "\n",
    "fasta_df['Extended Sequences'] = nucleotides_list\n",
    "fasta_df['Strand'] = strand_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c8c6151",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove gathered sequences that contain an intetermindant nucleotide \n",
    "\n",
    "fasta_df = fasta_df[~fasta_df['Extended Sequences'].str.contains('N')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69511d3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add on experimental sequences \n",
    "\n",
    "fasta_df['Twist Prepped Sequence'] =  (five_seq + fasta_df['Extended Sequences'].str[5:] + three_seq) # Remove first 5 nucleotides of leader as can interfere with aptamer structure and aptamer starts at position 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b00acd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate length \n",
    "\n",
    "fasta_df['Twist Length'] = fasta_df['Twist Prepped Sequence'].apply(len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dbe0585",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove duplicate oligos \n",
    "\n",
    "fasta_df = fasta_df.drop_duplicates(subset=['Extended Sequences'], keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc381e83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export prepared oligos to Excel \n",
    "\n",
    "fasta_df.to_excel('20240408_Twist_Order.xlsx',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64056711",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
